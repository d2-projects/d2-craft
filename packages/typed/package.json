{
  "name": "@d2-craft/typed",
  "version": "0.0.2"
}
