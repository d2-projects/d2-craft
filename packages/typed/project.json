{
  "root": "packages/typed",
  "sourceRoot": "packages/typed/src",
  "projectType": "library",
  "targets": {
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["packages/typed/**/*.ts"]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["coverage/packages/typed"],
      "options": {
        "jestConfig": "packages/typed/jest.config.js",
        "passWithNoTests": true
      }
    }
  },
  "tags": []
}
