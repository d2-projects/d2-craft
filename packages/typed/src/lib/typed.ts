export interface NodeMetaBase<Name, Config, _NodeMeta> {
  component: Name,
  config?: Config
  children?: _NodeMeta[]
}

export type NodeMeta<Ex = null> = Ex

/**
 * @example
 *
 *   type ExItem<N, C> = NodeMetaBase<N, C, NodeMeta<ExNodeMeta>>
 *   type ExNodeMeta<Ex = null> = NodeMeta< Ex
 *     | ExItem<'A', { a: string }>
 *     | ExItem<'B', { b: string }>
 *   >
 *
 *   const testTree: RootMeta<ExNodeMeta> = {
 *     children: [{
 *       component: 'A',
 *       config: { a: '' },
 *       children: [{
 *         component: 'B',
 *         config: { b: '' },
 *         children: [{
 *           component: 'A'
 *         }]
 *       }]
 *     }]
 *   }
 */
export interface RootMeta<ExtendableNodeMeta> {
  children: ExtendableNodeMeta[]
}
