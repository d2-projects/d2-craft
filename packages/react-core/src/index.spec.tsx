import { render } from '@testing-library/react';
import { ExUnused, NodeMeta, NodeMetaBase, RootMeta } from '@d2-craft/typed';
import { CraftProvider, CraftRoot, useCraftNode, makeComponentMap, CraftNode, useCraftProvider } from './index'
import React from 'react';

describe('React Core', () => {
  it('should render successfully', () => {

    // Example component A
    type ComponentMetaA = ExItem<'A', { a: string }>;

    const ComponentA: React.FC = () => {
      const { meta } = useCraftNode<ComponentMetaA>()
      const { componentMap } = useCraftProvider()

      return (
        <div className="component-a" data-craft-uid={meta.__uid}>
          <b>Component A: { meta.config?.a }</b>
          <div className="component-a-children">
            {
              meta.children?.map((child, key) => {
                const Component = componentMap.get(child.component)
                return Component && (
                  <CraftNode meta={child} key={key}>
                    <Component />
                  </CraftNode>
                )
              })
            }
          </div>
        </div>
      )
    }

    // Example component B
    type ComponentMetaB = ExItem<'B', { b: string }>

    const ComponentB: React.FC = () => {
      const { meta } = useCraftNode<ComponentMetaB>()
      const { componentMap } = useCraftProvider()

      return (
        <div className="component-b" data-craft-uid={meta.__uid}>
          <b>Component B: { meta.config?.b }</b>
          <div className="component-b-children">
            {
              meta.children?.map((child, key) => {
                const Component = componentMap.get(child.component)
                return Component && (
                  <CraftNode meta={child} key={key}>
                    <Component />
                  </CraftNode>
                )
              })
            }
          </div>
        </div>
      )
    }

    // Example component C
    type ComponentMetaC = ExItem<'C', { c: string }>

    const ComponentC: React.FC = () => {
      const { meta } = useCraftNode<ComponentMetaC>()

      return (
        <div className="component-c" data-craft-uid={meta.__uid}>
          <b>Component C: { meta.config?.c }</b>
        </div>
      )
    }

    // Recursive components meta
    type ExItem<N, C> = NodeMetaBase<N, C, NodeMeta<ExNodeMeta>>
    type ExNodeMeta<Ex = ExUnused> = NodeMeta< Ex
      | ComponentMetaA
      | ComponentMetaB
      | ComponentMetaC
    >

    // Components map
    const componentMap = makeComponentMap()
      .append<ComponentMetaA>('A', ComponentA)
      .append<ComponentMetaB>('B', ComponentB)
      .append<ComponentMetaC>('C', ComponentC)
      .value()

    // Example data
    const rootMeta: RootMeta<ExNodeMeta> = {
      children: [{
        component: 'A',
        config: { a: '1' },
        children: [{
          component: 'B',
          config: { b: '2' },
          children: [{
            component: 'C',
            config: { c: '3' }
          }]
        }]
      }]
    }

    // Three layer construct
    const { baseElement } = render(
      <CraftProvider componentMap={componentMap}>
        <CraftRoot<ExNodeMeta> meta={rootMeta}>
          {(rootMeta) => (
            <div className="craft-root">
              {rootMeta.children.map((child, key) => {
                const Component = componentMap.get(child.component)
                return Component && (
                  <CraftNode meta={child} key={key}>
                    <Component />
                  </CraftNode>
                )
              })}
            </div>
          )}
        </CraftRoot>
      </CraftProvider>
    );

    expect(baseElement).toBeTruthy();
    expect(baseElement).toMatchSnapshot()
  });
});
