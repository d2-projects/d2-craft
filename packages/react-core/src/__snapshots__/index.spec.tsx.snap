// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React Core should render container and children with item wrapper 1`] = `
<body>
  <div>
    <div
      class="craft-root"
    >
      <div
        class="flex-container"
        data-craft-uid="0"
      >
        <div
          class="flex-container-item"
          data-item-uid="1"
        >
          <div
            class="text"
            data-craft-uid="1"
          >
            <b>
              Hello
            </b>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`React Core should render default error boundary component 1`] = `
<body>
  <div>
    <div
      class="craft-root"
    >
      <div
        data-error-craft-uid="0"
      >
        RENDER ERROR
      </div>
    </div>
  </div>
</body>
`;

exports[`React Core should render successfully 1`] = `
<body>
  <div>
    <div
      class="craft-root"
    >
      <div
        class="component-a"
        data-craft-uid="0"
      >
        <b>
          Component A: 
          1
        </b>
        <div
          class="component-a-children"
        >
          <div
            class="component-b"
            data-craft-uid="1"
          >
            <b>
              Component B: 
              2
            </b>
            <div
              class="component-b-children"
            >
              <div
                class="component-c"
                data-craft-uid="2"
              >
                <b>
                  Component C: 
                  3
                </b>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
